### DISCUSSION TITLE
Migration and Support for MariaDB/MySQL Database

### DISCUSSION THREAD

**Discussion by homer-simpson** [ArchiveBox/ArchiveBox#1587] [Q&A]
**Background:**

Currently, the application relies on a SQLite database for data storage. While this works well for smaller datasets and lighter usage, SQLite’s performance begins to degrade as data volume grows, impacting overall application performance.

**Feature Suggestion:**

To improve performance and scalability, I propose adding support for MariaDB/MySQL as an alternative database backend. This feature should allow users to migrate seamlessly from SQLite to a MariaDB/MySQL instance without manual data transfer.

**Proposed Implementation:**

1. Automatic Migration: A migration script or command that would transfer all data from the existing SQLite database to a newly configured MariaDB/MySQL database. This script should:
- Be easy to run and configure.
- Retain all current data and structures from SQLite without data loss.

2. Configuration Options: Update the configuration to support database selection between SQLite and MariaDB/MySQL.

3. Documentation Update: Include detailed documentation to guide users through the setup and migration process for MariaDB/MySQL.

**Benefits:**

- Improved performance and stability for larger datasets.
- Enhanced scalability as data grows.
- Flexibility for users to select the database that best suits their needs.

This feature would significantly enhance usability for users who find that SQLite’s performance has reached its limits and would future-proof the application for growth.

**Comment by pirate** [comment]
What evidence / benchmarks do you have that show SQLite is actually the bottleneck?

> SQLite’s performance begins to degrade as data volume grows

I have not found this to be true. 

There are plenty of slow reads due to lack of indexes on some columns and legacy python code that loads everything into memory, but that has nothing to do with SQLite.

For more context see previous discussions about switching DBs:
- https://github.com/ArchiveBox/ArchiveBox/issues/601#issuecomment-817446169
- https://github.com/ArchiveBox/ArchiveBox/issues/787#issuecomment-878518747

**Reply by homer-simpson** [reply]
Okay - that may be true and you can certainly judge that better since you know the code.

I now have over 6500 snapshots in the archive box and it takes forever to load the last pages. See screenshot. It takes about 10 minutes for the page to load completely.

The snapshot is complete after about 30 seconds, the rest - especially the favicon files - arrive minutes later.

![image](https://github.com/user-attachments/assets/ca55d3d1-d68d-4f19-b451-037b83e04862)

Initially I had the archive box running in a Docker container on a QNap TVS-663 with 16GB RAM and have now moved it to better hardware (12th Gen Intel(R) Core(TM) i7-1265U (10 cores, 12 threads), 32GB RAM). However, this did not really lead to a noticeable improvement.

If I can help to analyze the problem in more detail and you can tell me what I need to do, I would be very happy to do so :-)

Thanks for the links!

**Reply by pirate** [reply]
What page is this you're trying to load? 

Also please share the full output of `archivebox version`.

**Reply by homer-simpson** [reply]
archivebox@1bd7372b95d3:~$ archivebox version
0.7.2
ArchiveBox v0.7.2 COMMIT_HASH=315c9f3 BUILD_TIME=2024-04-24 22:47:02 1713998822
IN_DOCKER=True IN_QEMU=False ARCH=x86_64 OS=Linux PLATFORM=Linux-6.11.6-2-default-x86_64-with-glibc2.36 PYTHON=Cpython
FS_ATOMIC=True FS_REMOTE=False FS_USER=999:999 FS_PERMS=644
DEBUG=False IS_TTY=True TZ=UTC SEARCH_BACKEND=ripgrep LDAP=False

[i] Dependency versions:
 √  PYTHON_BINARY         v3.11.9         valid     /usr/local/bin/python3.11                                                   
 √  SQLITE_BINARY         v2.6.0          valid     /usr/local/lib/python3.11/sqlite3/dbapi2.py                                 
 √  DJANGO_BINARY         v3.1.14         valid     /usr/local/lib/python3.11/site-packages/django/__init__.py                  
 √  ARCHIVEBOX_BINARY     v0.7.2          valid     /usr/local/bin/archivebox                                                   

 √  CURL_BINARY           v8.5.0          valid     /usr/bin/curl                                                               
 √  WGET_BINARY           v1.21.3         valid     /usr/bin/wget                                                               
 √  NODE_BINARY           v20.12.2        valid     /usr/bin/node                                                               
 √  SINGLEFILE_BINARY     v1.1.46         valid     /app/node_modules/single-file-cli/single-file                               
 √  READABILITY_BINARY    v0.0.11         valid     /app/node_modules/readability-extractor/readability-extractor               
 √  MERCURY_BINARY        v1.0.0          valid     /app/node_modules/@postlight/parser/cli.js                                  
 √  GIT_BINARY            v2.39.2         valid     /usr/bin/git                                                                
 √  YOUTUBEDL_BINARY      v2023.12.30     valid     /usr/local/bin/yt-dlp                                                       
 X  CHROME_BINARY         ?               invalid   chromium                                                                    
 √  RIPGREP_BINARY        v13.0.0         valid     /usr/bin/rg                                                                 

[i] Source-code locations:
 √  PACKAGE_DIR           23 files        valid     /app/archivebox                                                             
 √  TEMPLATES_DIR         3 files         valid     /app/archivebox/templates                                                   
 -  CUSTOM_TEMPLATES_DIR  -               disabled  None                                                                        

[i] Secrets locations:
 -  CHROME_USER_DATA_DIR  -               disabled  None                                                                        
 -  COOKIES_FILE          -               disabled  None                                                                        


[i] Data locations: (not in a data directory)

[!] Warning: Missing 1 recommended dependencies
    ! CHROME_BINARY: chromium (unable to detect version)
      Hint: To install all packages automatically run: archivebox setup
            or to disable it and silence this warning: archivebox config --set SAVE_CHROME=False

**Reply by homer-simpson** [reply]
This is my docker configuration

{
    "AppArmorProfile": "",
    "Args": [
        "--",
        "/app/bin/docker_entrypoint.sh",
        "archivebox",
        "server",
        "--quick-init",
        "0.0.0.0:8000"
    ],
    "Config": {
        "AttachStderr": false,
        "AttachStdin": false,
        "AttachStdout": false,
        "Cmd": [
            "archivebox",
            "server",
            "--quick-init",
            "0.0.0.0:8000"
        ],
        "Domainname": "",
        "Entrypoint": [
            "dumb-init",
            "--",
            "/app/bin/docker_entrypoint.sh"
        ],
        "Env": [
            "ARCHIVEBOX_USER=archivebox",
            "CHROME_BINARY=chromium",
            "CHROME_SANDBOX=False",
            "USE_MERCURY=True",
            "USE_READABILITY=True",
            "USE_SINGLEFILE=True",
            "SAVE_MEDIA=True",
            "SAVE_PDF=True",
            "SAVE_READABILITY=True",
            "SAVE_SCREENSHOT=True",
            "SAVE_SINGLEFILE=True",
            "PATH=/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/venv/bin:/app/.venv/bin:/app/node_modules/.bin",
            "LANG=C.UTF-8",
            "GPG_KEY=*DONOTSHOW*",
            "PYTHON_VERSION=3.11",
            "PYTHON_PIP_VERSION=24.0",
            "PYTHON_SETUPTOOLS_VERSION=65.5.1",
            "PYTHON_GET_PIP_URL=https://github.com/pypa/get-pip/raw/dbf0c85f76fb6e1ab42aa672ffca6f0a675d9ee4/public/get-pip.py",
            "PYTHON_GET_PIP_SHA256=dfe9fd5c28dc98b5ac17979a953ea550cec37ae1b47a5116007395bfacff2ab9",
            "TZ=UTC",
            "LANGUAGE=en_US:en",
            "LC_ALL=C.UTF-8",
            "DEBIAN_FRONTEND=noninteractive",
            "APT_KEY_DONT_WARN_ON_DANGEROUS_USAGE=1",
            "PYTHONIOENCODING=UTF-8",
            "PYTHONUNBUFFERED=1",
            "PIP_DISABLE_PIP_VERSION_CHECK=1",
            "npm_config_loglevel=error",
            "NODE_VERSION=20",
            "DEFAULT_PUID=911",
            "DEFAULT_PGID=911",
            "CODE_DIR=/app",
            "DATA_DIR=/data",
            "GLOBAL_VENV=/venv",
            "PLAYWRIGHT_BROWSERS_PATH=/browsers",
            "APP_VENV=/app/.venv",
            "NODE_MODULES=/app/node_modules",
            "IN_DOCKER=True"
        ],
        "ExposedPorts": {
            "8000/tcp": {}
        },
        "Hostname": "1bd7372b95d3",
        "Image": "docker.io/archivebox/archivebox:latest",
        "Labels": {
            "description": "All-in-one personal internet archiving container",
            "documentation": "https://github.com/ArchiveBox/ArchiveBox/wiki/Docker#docker",
            "homepage": "https://github.com/ArchiveBox/ArchiveBox",
            "maintainer": "Nick Sweeting <dockerfile@archivebox.io>",
            "name": "archivebox"
        },
        "OnBuild": null,
        "OpenStdin": false,
        "StdinOnce": false,
        "Tty": false,
        "User": "",
        "Volumes": {
            "/data": {}
        },
        "WorkingDir": "/data"
    },
    "Created": "2024-11-08T11:42:38.036916823Z",
    "Driver": "overlay2",
    "ExecIDs": [
        "c48d32194e1d8ec2424b69f6f46ce760aa5cd017fd6c3c2588d07cdaaa2bda7e"
    ],
    "GraphDriver": {
        "Data": {
            "LowerDir": "/var/lib/docker/overlay2/b2fac8feb5a15ad63b3398495affc248b5886b050c48a4cd1637b5c0175bb3c3-init/diff:/var/lib/docker/overlay2/e478e82bf17e5fc52bf2c1a64c96d1e49696d056e225e2bbcbed613f4eb6723a/diff:/var/lib/docker/overlay2/293e52d203dc726d7c90a41095ff4d9180f7ac383b48c3e5fdef1fdc1a161310/diff:/var/lib/docker/overlay2/9dc90c4a84a7b8c11c78627a404a15c3a23362441539a06579eb846776c2028b/diff:/var/lib/docker/overlay2/ec5b6200a8e4f7c6f170a5d849b3bea72966842f16c928455ad1f220c86b211e/diff:/var/lib/docker/overlay2/17d642cdf12e3f7299649991970e0636f356821b4a4eba77b58bd93e9dec29fd/diff:/var/lib/docker/overlay2/37a7d5549196cfe873a68ff36c5ccc5c59c045017cfbbd0b7053e7368510abd2/diff:/var/lib/docker/overlay2/226a8fb5f33d9813f5286bbc7e1a93a79f0a09976437cfdf4256331e4a6e802c/diff:/var/lib/docker/overlay2/75608f1df3eab451d17772eb5f23060c42c2bad8ac097c688874de3ec40d756c/diff:/var/lib/docker/overlay2/db4d6a204e9573156532463222ba1ea9a434e790f22823c4f31fa7f088f5e4e1/diff:/var/lib/docker/overlay2/de230a1793bd02cf7ce4ebaef701ccde140797b3323b75416d17ffec97f55fd6/diff:/var/lib/docker/overlay2/0b44606e8ddb958b334fe9eb0438ef6496266a03cab967f76d409ff3a4bc3091/diff:/var/lib/docker/overlay2/d0e1d9d4e2e9d29ae7442152e2490ba99f03a092163650da7a3c33695aeb39a1/diff:/var/lib/docker/overlay2/379c3e8c2afe12bc6237edb94f4b1d20ac7d2b12fdaca9b690b0ef4e483e1fc1/diff:/var/lib/docker/overlay2/f4e09660fc086681bfa2dc61aeddc4540736718027faa00962ee37f79164c8e3/diff:/var/lib/docker/overlay2/c934023cd1003720f5ba6855b109cb1ce4990eac3698d2f42d5336d3a843fbe6/diff:/var/lib/docker/overlay2/0c880f2cef34378d5d53c8be1b9e6f7a9292d36f97315f50ef922fb0d9c8d227/diff:/var/lib/docker/overlay2/79ba867b3e0b10185407921080f45d80c01205250592b7991d1d6ba383dccacc/diff:/var/lib/docker/overlay2/2307ec827aaaebec7de8bf82a28fbe8fa99b025cbc5caba1a9744ef52ff51e44/diff:/var/lib/docker/overlay2/714661baa128807d77990180ddc079faf7463ed0173402dd30052314193d6022/diff:/var/lib/docker/overlay2/1aba086dda586121d4a7f0a8b0f78fd146796a06ca59e7ed17306ed335335005/diff:/var/lib/docker/overlay2/74a748130b2bcd303956fc2de0e82d0616e0f6db24b69bfc489d9b74d0fd1a78/diff:/var/lib/docker/overlay2/41018aeff8f30c5ca1afa14f5550e0a0b3e0c41ee8480c47e36b59c1a5651f1c/diff:/var/lib/docker/overlay2/55ec1f7b43d455033260a97c6226ce6f1c406af880a1b0380c307873a584c5fd/diff:/var/lib/docker/overlay2/7fe4b14b59013a93b31f64c74408dacdc36688775662e2aeb0f0ec1408d5602b/diff:/var/lib/docker/overlay2/cb37446564d47a921d79aee216ce87405e4f3ff3b4fd285d5a4c6c556d9ee781/diff:/var/lib/docker/overlay2/4a0c589eb7d90c7ca077188fee00807745025a2386e64ef38663cd754959b2c9/diff:/var/lib/docker/overlay2/9c08dadcf1c059a931eed425037c1fa7c4fea8f18ddfaeb26f13385ef95beb8b/diff",
            "MergedDir": "/var/lib/docker/overlay2/b2fac8feb5a15ad63b3398495affc248b5886b050c48a4cd1637b5c0175bb3c3/merged",
            "UpperDir": "/var/lib/docker/overlay2/b2fac8feb5a15ad63b3398495affc248b5886b050c48a4cd1637b5c0175bb3c3/diff",
            "WorkDir": "/var/lib/docker/overlay2/b2fac8feb5a15ad63b3398495affc248b5886b050c48a4cd1637b5c0175bb3c3/work"
        },
        "Name": "overlay2"
    },
    "HostConfig": {
        "AutoRemove": false,
        "Binds": [
            "archivebox_data:/data"
        ],
        "BlkioDeviceReadBps": null,
        "BlkioDeviceReadIOps": null,
        "BlkioDeviceWriteBps": null,
        "BlkioDeviceWriteIOps": null,
        "BlkioWeight": 0,
        "BlkioWeightDevice": null,
        "CapAdd": [
            "AUDIT_WRITE",
            "CHOWN",
            "DAC_OVERRIDE",
            "FOWNER",
            "FSETID",
            "KILL",
            "MKNOD",
            "NET_BIND_SERVICE",
            "NET_RAW",
            "SETFCAP",
            "SETGID",
            "SETPCAP",
            "SETUID",
            "SYS_CHROOT"
        ],
        "CapDrop": [
            "AUDIT_CONTROL",
            "BLOCK_SUSPEND",
            "DAC_READ_SEARCH",
            "IPC_LOCK",
            "IPC_OWNER",
            "LEASE",
            "LINUX_IMMUTABLE",
            "MAC_ADMIN",
            "MAC_OVERRIDE",
            "NET_ADMIN",
            "NET_BROADCAST",
            "SYSLOG",
            "SYS_ADMIN",
            "SYS_BOOT",
            "SYS_MODULE",
            "SYS_NICE",
            "SYS_PACCT",
            "SYS_PTRACE",
            "SYS_RAWIO",
            "SYS_RESOURCE",
            "SYS_TIME",
            "SYS_TTY_CONFIG",
            "WAKE_ALARM"
        ],
        "Cgroup": "",
        "CgroupParent": "",
        "CgroupnsMode": "private",
        "ConsoleSize": [
            0,
            0
        ],
        "ContainerIDFile": "",
        "CpuCount": 0,
        "CpuPercent": 0,
        "CpuPeriod": 0,
        "CpuQuota": 0,
        "CpuRealtimePeriod": 0,
        "CpuRealtimeRuntime": 0,
        "CpuShares": 0,
        "CpusetCpus": "",
        "CpusetMems": "",
        "DeviceCgroupRules": null,
        "DeviceRequests": [],
        "Devices": [],
        "Dns": [],
        "DnsOptions": [],
        "DnsSearch": [],
        "ExtraHosts": [],
        "GroupAdd": null,
        "IOMaximumBandwidth": 0,
        "IOMaximumIOps": 0,
        "Init": false,
        "IpcMode": "private",
        "Isolation": "",
        "Links": null,
        "LogConfig": {
            "Config": {
                "max-file": "5",
                "max-size": "10m"
            },
            "Type": "json-file"
        },
        "MaskedPaths": [
            "/proc/asound",
            "/proc/acpi",
            "/proc/kcore",
            "/proc/keys",
            "/proc/latency_stats",
            "/proc/timer_list",
            "/proc/timer_stats",
            "/proc/sched_debug",
            "/proc/scsi",
            "/sys/firmware",
            "/sys/devices/virtual/powercap"
        ],
        "Memory": 0,
        "MemoryReservation": 0,
        "MemorySwap": 0,
        "MemorySwappiness": null,
        "NanoCpus": 0,
        "NetworkMode": "bridge",
        "OomKillDisable": null,
        "OomScoreAdj": 0,
        "PidMode": "",
        "PidsLimit": null,
        "PortBindings": {
            "8000/tcp": [
                {
                    "HostIp": "",
                    "HostPort": "11015"
                }
            ]
        },
        "Privileged": false,
        "PublishAllPorts": false,
        "ReadonlyPaths": [
            "/proc/bus",
            "/proc/fs",
            "/proc/irq",
            "/proc/sys",
            "/proc/sysrq-trigger"
        ],
        "ReadonlyRootfs": false,
        "RestartPolicy": {
            "MaximumRetryCount": 0,
            "Name": "always"
        },
        "Runtime": "runc",
        "SecurityOpt": null,
        "ShmSize": 67108864,
        "UTSMode": "",
        "Ulimits": null,
        "UsernsMode": "",
        "VolumeDriver": "",
        "VolumesFrom": null
    },
    "HostnamePath": "/var/lib/docker/containers/1bd7372b95d3ba964ffbe46afce4697aec55d581fb77f1b93f23550df877f287/hostname",
    "HostsPath": "/var/lib/docker/containers/1bd7372b95d3ba964ffbe46afce4697aec55d581fb77f1b93f23550df877f287/hosts",
    "Id": "1bd7372b95d3ba964ffbe46afce4697aec55d581fb77f1b93f23550df877f287",
    "Image": "sha256:b50ff8d34a8201b0da7aef68ecd4718cfe9e54a649f7ae4faf1cc3fbc0274bf3",
    "LogPath": "/var/lib/docker/containers/1bd7372b95d3ba964ffbe46afce4697aec55d581fb77f1b93f23550df877f287/1bd7372b95d3ba964ffbe46afce4697aec55d581fb77f1b93f23550df877f287-json.log",
    "MountLabel": "",
    "Mounts": [
        {
            "Destination": "/data",
            "Driver": "local",
            "Mode": "z",
            "Name": "archivebox_data",
            "Propagation": "",
            "RW": true,
            "Source": "/var/lib/docker/volumes/archivebox_data/_data",
            "Type": "volume"
        }
    ],
    "Name": "/archivebox",
    "NetworkSettings": {
        "Bridge": "",
        "EndpointID": "",
        "Gateway": "",
        "GlobalIPv6Address": "",
        "GlobalIPv6PrefixLen": 0,
        "HairpinMode": false,
        "IPAddress": "",
        "IPPrefixLen": 0,
        "IPv6Gateway": "",
        "LinkLocalIPv6Address": "",
        "LinkLocalIPv6PrefixLen": 0,
        "MacAddress": "",
        "Networks": {
            "bridge": {
                "Aliases": [
                    "1bd7372b95d3"
                ],
                "DNSNames": [
                    "archivebox",
                    "1bd7372b95d3"
                ],
                "DriverOpts": null,
                "EndpointID": "6f21e2a29dc75409554aa27fb0afe9b828671111800399ed152aea5d7b3dd7a3",
                "Gateway": "172.22.10.1",
                "GlobalIPv6Address": "",
                "GlobalIPv6PrefixLen": 0,
                "IPAMConfig": {},
                "IPAddress": "172.22.10.5",
                "IPPrefixLen": 24,
                "IPv6Gateway": "",
                "Links": null,
                "MacAddress": "02:42:ac:16:0a:05",
                "NetworkID": "b3967447f2f26df4ada9a4aa82328ab38f85537f69f0aeca3b2422d09d1e03b3"
            }
        },
        "Ports": {
            "8000/tcp": [
                {
                    "HostIp": "0.0.0.0",
                    "HostPort": "11015"
                },
                {
                    "HostIp": "::",
                    "HostPort": "11015"
                }
            ]
        },
        "SandboxID": "5d692bceaafa94057091ab799e46244b40a2f3c05e6e24d99b08459d872e93f5",
        "SandboxKey": "/var/run/docker/netns/5d692bceaafa",
        "SecondaryIPAddresses": null,
        "SecondaryIPv6Addresses": null
    },
    "Path": "dumb-init",
    "Platform": "linux",
    "Portainer": {
        "ResourceControl": {
            "Id": 69,
            "ResourceId": "1bd7372b95d3ba964ffbe46afce4697aec55d581fb77f1b93f23550df877f287",
            "SubResourceIds": [],
            "Type": 1,
            "UserAccesses": [],
            "TeamAccesses": [],
            "Public": false,
            "AdministratorsOnly": true,
            "System": false
        }
    },
    "ProcessLabel": "",
    "ResolvConfPath": "/var/lib/docker/containers/1bd7372b95d3ba964ffbe46afce4697aec55d581fb77f1b93f23550df877f287/resolv.conf",
    "RestartCount": 0,
    "State": {
        "Dead": false,
        "Error": "",
        "ExitCode": 0,
        "FinishedAt": "2024-11-09T04:29:42.340494488Z",
        "OOMKilled": false,
        "Paused": false,
        "Pid": 2460,
        "Restarting": false,
        "Running": true,
        "StartedAt": "2024-11-09T05:30:26.129859716Z",
        "Status": "running"
    }
}

**Comment by 2ndLtAmazing** [comment]
For what it's worth, I really like the idea of using MySQL/MariaDB.  At the very least, it would be possible to have the database living on entirely different hardware, and that would make troubleshooting performance easier. I'm currently trying to troubleshoot some pretty serious performance issues with my install, and it would be great if I could more easily isolate the database as a culprit in this process.

Not to mention opening the door for zero-effort manual queries of the database, and/or building out separate independent utilities that could read it.

Another thing is that as was mentioned there have been observed performance issues from a lack of indexes on certain columns, it would be seamless to add indexes to the MySQL/MariaDB without any other part of the code needing to be updated or even aware of it.  If such a thing were discovered it would be trivial for somebody to update index structure into the code.  Testing such a thing in SQLite, at the very least seems more cumbersome, and certainly accessible to less would-be drive-by problem spotters, if that makes any sense.

I hate to jump on a bandwagon of let's make the devs jump through ridiculous hoops so that the some users experience a 0.001% convenience improvement, but I would like to think that adding an option to utilize a defined MySQL/MariaDB source wouldn't be a ridiculously complicated ordeal.  (I could be wrong, of course.  I'm just thinking that if the tables and structure are the same, it wouldn't take a ton of refactoring.)

**Comment by pirate** [comment]
I'm going to close this for now as I have no plans to implement MySQL/MariaDB support, I think SQLite is better suited for long-term durability of data on disk. I'm open to supporting PostgreSQL someday but I definitely don't want to support multiple DBs beyond those two.

Unfortunately none of these are drop-in replacements for each other and every new DB requires quite a bit of work to smooth over the differences in UUID handling, JSON query support, concurrency settings, etc.
